### ISSUE TITLE
Draw bi-directional support

### ISSUE THREAD

**Issue by bhaskarvk** [bhaskarvk/leaflet.extras#17]
Currently with the draw plugin, you can

- Create/Resize/Move/Delete features.
- For each of the above activity you'll get Shiny events to handle data on R side.

But...

- There is no way to specify an initial set of features to work on.
- There is no way to programmatically add/update/delete features on Shiny side.

How critical is it to have these features ? Looking for input from the usual suspects.

**Comment by bhaskarvk** [comment]
Looking for inputs from @tim-salabim , @mdsumner, @timelyportfolio

**Comment by byzheng** [comment]
You can intitial feature in the drawItems. See https://github.com/rstudio/leaflet/issues/327. I just don't know how to add current structure of leaflet or leaflet.extras. I am looking this function too.

I think it is possible to programmatically edit/remove/add in shiny. This is the reason I want to add the layer ID in the shiny events. We can use layer ID to change feature. The alternative way could be add name to drawItems, then find layers by name.

**Comment by mdsumner** [comment]
Being able to have an initial set of features is critical, but the add/update/delete much less so, because you could always (re-)initialize the view with a new set.

**Comment by tim-salabim** [comment]
I totally agree w Mike, an initial set of features is critical. However, I do think that editing existing point/vertices is just as important. Of course in many situations we could simply modify the feature set on the R side programmatically and then re-initialise the view, however, there are many situations where visual manual modification is crucial. This is exactly what led me to create our RConsortium application:

https://github.com/environmentalinformatics-marburg/mapview_toolchain/blob/master/mapview_interactive_data_manipulation.Rmd

@timelyportfolio will soon start on this. The general idea is to create a generalised framework for passing things back into R workspace from JS widget-side (without shiny), because R is where we work, JS only provides the vis/interactive part. This part will basically be a "reverse **htmlwidgets**" and ideally not be limited to leaflet but usable for any htmlwidgets-based vis. Once we have this, we can start thinking about tools that are relevant for spatial data manipulation. If leaflet.draw doesn't allow manipulation of existing features, then we would likely need to find or implement something else. 

The biggest problem I have is that I have a heap of ideas and want-to-have features for interactive spatial visualisation and manipulation but have no way of implementing them as I do not speak JS well enough to even try. But that's to be discussed elsewhere...

To sum up my points, having leaflet.draw provides a nice little playground, but to be useful it would need to allow editing of existing features and this manipulated set needs to be passable back to R workspace.

**Comment by mdsumner** [comment]
I agree with @tim-salabim  that it's the input from the browser that is most important, and I was assuming that. I meant that *R* in the background being able to add/update/delete the content in the browser, a) after it was initialized b) after it had been edited/added-to/deleted by user interaction -  was less important - though now I wonder if I misunderstood? 

a) R initializes the map, either empty or with an existing vector layer
b) user draws into the layer any combination of - new features, edits features, removes features
c) user closes interactive session, and the results of b have replaced the content of a), in R

I read "There is no way to programmatically add/update/delete features on Shiny side." as being mechanisms of some other kind that could modify the contents in b) - i.e. the database gets updated, new data comes in, another widget in page emits updates (i.e. a scaling factor entered that applies only to "brushed" or otherwise selected features). 

Hopefully that all makes sense.

**Comment by bhaskarvk** [comment]
Thanks for the input guys.

a) Getting data from R to JS side is never going to be a problem. I think you guys misread when I said `There is no way to specify an initial set of features to work on.` & `There is no way to programmatically add/update/delete features on Shiny side.`
I was just pointing out that I haven't coded that part yet. That's relatively straight forward and I can code that up fairly easily.

b) If you have a Shiny app, then the whole bi-directional problem is solved. You can send data to JS, start draw with that existing data, do your manipulations, and read the modified geometries in R. And you can do this round-and-round. i.e. once the data is back in R modify it and send it back to JS for viz or further editing via draw.

c) If you want to do b) w/o Shiny, that's where things get trickier. Because w/o a server component on R side, I see no way to get back data from JS side. So maybe not Shiny but will need some kind of a server component on R side to maintain bi-directional communication channel.
w/o that it's just a one way traffic from R to JS. Perhaps @timelyportfolio can add his perspective.

So to summarize, bootstrapping draw with existing geometries is not a problem at all, I just need to code it. With shiny you can do pretty much any workflow you can think of. w/o shiny, getting data back and forth between R and JS will require some other communication channel between R and JS. 

I will go ahead and code up the a) part and b) part is just shiny as usual. For the c) part I think that's a problem outside of draw, heck even leaflet. You need to solve that at htmlwidgets level. Rest assured that draw will allow you to manipulate existing geometries.

**Comment by mdsumner** [comment]
Oh!  Sorry, yes I had totally misread it. 

This is so great!  👍 👍

**Comment by tim-salabim** [comment]
Hehe, yeah I also misunderstood! That's great news! If a) is available then people can implement individual solutions with b). Work on c) will start shortly and likely involve input from @ramnathv and @jcheng5 but I think @timelyportfolio has already been in contact with them regarding this.

**Comment by bhaskarvk** [comment]
On careful examination of Leaflet code (Both R and JS), this is not straightforward. It is doable but not as easy as I had anticipated.
Initially I had thought I should be able to support

```r
addPolygons(..., group='draw') %>%
addCircle(..., group='draw') %>%
addGeojson(..., group='draw') %>%
addDrawToolbar(..., group='draw', preserveExisting=TRUE)
```

So basically call any add* method specifying a group name, and pass that same group name to addDrawToolbar and set preserveExisting=TRUE. But...
When you specify a group name the Paths/GeoJSON is added to a `L.LayerGroup`, but draw requires a `L.FeatureGroup`, FeatureGroup inherits from LayerGroup and as such can be easily used where ever LayerGroup is being used.
But still it means changing Leaflet and adding stuff to FeatureGroup and not LayerGroup. I'll discuss with @jcheng5 and let you guys know.

**Comment by bhaskarvk** [comment]
OK I think I've got it.

`devtools::install_github('bhaskarvk/leaflet@develop')`  Until I push changes to rstudio master.

`devtools::install_github('bhaskarvk/leaflet.extras@feature/draw')`

And test case
http://rpubs.com/bhaskarvk/leaflet-draw 2nd example. It should work for all Paths/Markers. For now it won't work for GeoJSON/TopoJSON data, but I should be able to fix that too.

Could you guys test it out a bit and let me know, then I can push changes to rstudio/leaflet as well as leaflet.extras master.

Thanks

**Comment by bhaskarvk** [comment]
Btw this is should be compatible with adding new data programmatically via shiny's 

```r
leafletProxy() %>%
add*
```

As long as you use the same groupname.

**Comment by jcheng5** [comment]
@tim-salabim, @timelyportfolio, there's nothing in your R Consortium application that Shiny is not *perfectly* suited for. I've been waiting to talk about it but nobody has contacted me. Shiny is a generalized framework for passing data back and forth between JS and R, we've just added a UI toolkit on the client side and reactive programming framework on the server side, both of which you're free to use or not use (if you have any doubt about that, let me know and I can show you how each of these pieces relate to each other and how you can leave them behind).

You should create an alternative to Shiny if your goal is creating an alternative to Shiny. But if your goal is to make interactive JS widgets that have bidirectional communication with R, then just use Shiny--and if Shiny actually does not meet your needs for some reason, let's talk about it! (The main objection I can imagine is that Shiny apps block the console when you run them--but in the use cases I saw in the R Consortium application, I think that's what you want)

**Comment by tim-salabim** [comment]
@jcheng5 the reason for not having contacted you so far is slow negotiations between University of Marburg and the RConsortium. Felt like TTIP negotiations to me... This is all settled now and we are ready to sign the contract as far as I understand the situation, so work can start soon. If shiny is suited for this kind of communication we will surely use it, why reinvent the wheel. The issue of locking the console will need careful consideration as I can see situations where you'd want check something in the console while working on features in the browser. 
We will be in contact about all this as soon as we have the paperwork part sorted.

**Comment by bhaskarvk** [comment]
OK, I've added support to edit existing features either added via add<Markers/Polylines/Polygons> method or addGeoJSONv2 (also works with other add<FileType> methods of leaflet.extras).

The only issue is if the existing polygon/polyline geometry is complex (any real world polygon/line data), then the editing feature is very very slow. This is expected as the draw plugin has to create lots and lots of vertices. This can't be helped, but can be improved by simplifying shapes before adding to leaflet map.

Waiting on @jcheng5 to merge rstudio/leaflet#329, until then this stays in the feature/draw branch.

**Comment by tim-salabim** [comment]
Awesome! What about addCircleMarkers?

**Comment by bhaskarvk** [comment]
yes those too.

**Comment by bhaskarvk** [comment]
All fixed in d960816. A big warning editing complex geometries is very very slow, like it will at times hang your browser. Make sure to simplify polygons before adding them for editing.

**Comment by mdsumner** [comment]
Probably we will need topojson to improve this.

**Comment by tim-salabim** [comment]
Or tiled views

**Comment by mdsumner** [comment]
Implying demand-driven read from R?  Yes :)

**Comment by bhaskarvk** [comment]
Unless Leaflet natively supports topology, I don't see how topojson will help. We can currently add topojson to leaflet but it's converted to a geojson internally. Tiled views may help when you have lots of polygons.
The basic problem is that even a rudimentary state/county level polygon is complex enough that you need literally hundreds or even thousands of vertices to be able to edit them. So the only real solution is to have simplified polygons. I'll give you guys some sample code to play with later today.

**Comment by mdsumner** [comment]
Thanks @bhaskarvk appreciate this, it's helpful to get this kind of detail - it's hard to know what to ask for directly and so responses like this are really valuable!

**Comment by tim-salabim** [comment]
Does leaflet.draw have other limitations than leaflet itself? A few thousand points/vertices with leaflet is no problem.

**Comment by mdsumner** [comment]
I just tried with a typical "box model" for the Atlantis ecosystem model, and realized it's a good set of available layers for testing the editing: 

```R
#devtools::install_github('bhaskarvk/leaflet@develop')
#devtools::install_github('bhaskarvk/leaflet.extras@feature/draw')

#install.packages("rbgm")

library(rbgm)
set.seed(2)
## pick one of the available model files
fs <- sample(bgmfiles::bgmfiles(), 1)

## read the model, and convert to Spatial (box for polygons, face for boundary lines)
model <- boxSpatial(bgmfile(fs))
## most of the BGM models will be in a local map projection
model <- spTransform(model, "+init=epsg:4326")
library(geojsonio)

gs <- geojsonio::geojson_json(model)
#ts <- geojsonio::topojson_read(gs)

library(leaflet)
library(leaflet.extras)
leaflet() %>% addTiles() %>% addGeoJSONv2(gs, layerId = "model") %>%
  addDrawToolbar(
    targetLayerId='model',
    editOptions = editToolbarOptions(selectedPathOptions = selectedPathOptions())) %>%
  addLayersControl(overlayGroups = c('model'), options =
                     layersControlOptions(collapsed=FALSE)) %>%
  addStyleEditor()

```

There's a set of real-world polygon layers in the bgmfiles package, readable with the rbgm package, and you can get SpatialPolygonsDF with boxSpatial, lines with faceSpatial.

**Comment by bhaskarvk** [comment]
You don't need to build from those branches anymore, everything is now in trunk so
```r
devtools::install_github('rstudio/leaflet')
devtools::install_github('bhaskarvk/leaflet.extras')
```

**Comment by bhaskarvk** [comment]
Also you don't need to convert the SPDF to geojson, draw supports editing GeoJSON layers as well as shapes added to groups.

```r
model <- spTransform(model, "+init=epsg:4326")

leaflet() %>% addTiles() %>%
  addPolygons(data = model, group = 'model') %>%
  addDrawToolbar(targetGroup = 'model',
    editOptions = editToolbarOptions(
      selectedPathOptions = selectedPathOptions())) %>%
  addLayersControl(overlayGroups = c('model'), options =
                     layersControlOptions(collapsed=FALSE)) %>%
  addStyleEditor()

```

This dataset is fast enough. I didn't see any lag when enabling editing of these polygons.

**Comment by bhaskarvk** [comment]
Try this for performance comparison

```r
fName <- 'https://rawgit.com/benbalter/dc-maps/master/maps/ward-2012.geojson'

geoJson <- readr::read_file(fName)
geoJson2 <- rmapshaper::ms_simplify(geoJson, keep=0.01)

# Simplified geometry, everything works.
leaflet() %>% addTiles() %>% setView(-77.0369, 38.9072, 11) %>%
  addGeoJSONv2(geoJson2,
    group = 'wards', layerId = 'dc-wards') %>%
  addDrawToolbar(
    targetLayerId='dc-wards',
    editOptions = editToolbarOptions(
      selectedPathOptions = selectedPathOptions()))  %>%
  addLayersControl(overlayGroups = c('wards'),
                   options = layersControlOptions(collapsed=FALSE)) %>%
  addStyleEditor()

## Complex geometry clicking on the edit button will hang browser
leaflet() %>% addTiles() %>% setView(-77.0369, 38.9072, 11) %>%
  addGeoJSONv2(geoJson,
    group = 'wards', layerId = 'dc-wards') %>%
  addDrawToolbar(
    targetLayerId='dc-wards',
    editOptions = editToolbarOptions(
      selectedPathOptions = selectedPathOptions()))  %>%
  addLayersControl(overlayGroups = c('wards'),
                   options = layersControlOptions(collapsed=FALSE)) %>%
  addStyleEditor()
```

I tried various params for the `keep` option and even the default 0.05 is slow for editing, only when I got to 0.01 did I get the desired performance.

**Comment by bhaskarvk** [comment]
OK I think with shape simplification the performance is quite acceptable. And @tim-salabim indeed this is a draw plugin limitation, not a leaflet limitation.
You can check all the sample code  http://rpubs.com/bhaskarvk/draw.
